File: Day22.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <array>
#include <vector>
#include <algorithm>

using i64 = long long;
#define ALL(x) (x).begin(),(x).end()

struct Region {
    Region() = default;
    Region(int a, int b, int c, int d, int e, int f) : minX(a), maxX(b), minY(c), maxY(d), minZ(e), maxZ(f) {}
    i64 Volume() const { return (maxX + 1ll - minX) * (maxY + 1ll - minY) * (maxZ + 1ll - minZ); }
    constexpr bool Intersects(const Region& r)
    {
        if (maxX < r.minX || minX > r.maxX ||
            maxY < r.minY || minY > r.maxY ||
            maxZ < r.minZ || minZ > r.maxZ)
            return false;
        return true;
    }
    Region Intersection(const Region& r) const
    {
        std::array<int, 4> work{ minX, maxX, r.minX, r.maxX };
        std::sort(ALL(work));
        int x = work[1], X = work[2];
        work = { minY, maxY, r.minY, r.maxY };
        std::sort(ALL(work));
        int y = work[1], Y = work[2];
        work = { minZ, maxZ, r.minZ, r.maxZ };
        std::sort(ALL(work));
        return Region(x, X, y, Y, work[1], work[2]);
    }
    int minX, maxX, minY, maxY, minZ, maxZ;
};

int main(int argc, char* argv[])
{
    if (argc < 2)
    {
        std::cout << "Day19.exe inputfilename" << std::endl;
        return -1;
    }

    std::ifstream in(argv[1], std::ios::in);
    if (!in)
    {
        std::cout << "Could not open " << argv[1] << std::endl;
        return -1;
    }

    std::string s;
    i64 part1 = 0, part2 = 0;
    std::vector<Region> toAdd, toRemove;
    while (in >> s)
    {
        char c;
        int minX, maxX, minY, maxY, minZ, maxZ;
        in >> c >> c >> minX >> c >> c >> maxX >> c >> c >> c >> minY >> c >> c>> maxY;
        in >> c >> c >> c >> minZ >> c >> c >> maxZ;

        if (!part1 && (minX > 50 || maxX < -50 || minY >50 || maxY < -50 || minZ > 50 || maxZ < -50))
        {
            for (const Region& r : toAdd)
                part1 += r.Volume();
            for (const Region& r : toRemove)
                part1 -= r.Volume();
        }
        Region region(minX, maxX, minY, maxY, minZ, maxZ);
        std::vector<Region> work;
        if (s == "on")
        {
            for (const Region& r : toAdd)
                if (region.Intersects(r))
                    work.emplace_back(region.Intersection(r));
            toAdd.push_back(region);
            for (const Region& r : toRemove)
                if (region.Intersects(r))
                    toAdd.emplace_back(region.Intersection(r));
            toRemove.insert(toRemove.end(), ALL(work));
        }
        else
        {
            for (const Region& r : toRemove)
                if (region.Intersects(r))
                    work.push_back(region.Intersection(r));
            for (const Region& r : toAdd)
                if (region.Intersects(r))
                    toRemove.emplace_back(region.Intersection(r));
            toAdd.insert(toAdd.end(), ALL(work));
        }
    }

    for (const Region& r : toAdd)
        part2 += r.Volume(); 
    for (const Region& r : toRemove)
        part2 -= r.Volume();
    
    std::cout << "Part 1: " << part1 << "\nPart 2: " << part2 << std::endl;
    return 0;
}

File: Day22.fs
open System.IO

let Process (s:string) =
    let data = s.Split(' ')
    let simple = data.[1].Replace("..", ".").Replace("x=", "").Replace(",y=", ".").Replace(",z=",".")
    let bounds = simple.Split('.') |> Array.map System.Int64.Parse
    (data.[0].[1]='n', (bounds.[0], bounds.[1], bounds.[2], bounds.[3], bounds.[4], bounds.[5]))

let IsInit (_,(x,X,y,Y,z,Z)) = 
    (Seq.filter(fun x -> 50L < abs x) [x;X;y;Y;z;Z] |> Seq.length)=0

let Intersects (x0,X0,y0,Y0,z0,Z0) (x1,X1,y1,Y1,z1,Z1) =
    not (X0<x1 || X1<x0 || Y0<y1 || Y1<y0 || Z0<z1 || Z1<z0)

let Intersection (x0,X0,y0,Y0,z0,Z0) (x1,X1,y1,Y1,z1,Z1) =
    let x = Array.sort [|x0; X0; x1; X1|]
    let y = Array.sort [|y0; Y0; y1; Y1|]
    let z = Array.sort [|z0; Z0; z1; Z1|];
    (x.[1], x.[2], y.[1], y.[2], z.[1], z.[2])

let Volume (x,X,y,Y,z,Z) = (X-x+1L)*(Y-y+1L)*(Z-z+1L)

let Activate (inst:(bool*'a)[]) =
    let mutable toAdd, toRemove = [||], [||]
    Array.iter(fun (on, x) ->
        let mutable work = [||]
        let AppendWork x = work<-Array.append work [|x|]
        if on then
            Array.iter(fun y -> if Intersects x y then AppendWork (Intersection x y)) toAdd
            toAdd<-Array.append toAdd [|x|]
            Array.iter(fun y -> if Intersects x y then toAdd <- Array.append toAdd [|Intersection x y|]) toRemove
            toRemove<-Array.append toRemove work                
        else
            Array.iter(fun y -> if Intersects x y then AppendWork (Intersection x y)) toRemove
            Array.iter(fun y -> if Intersects x y then toRemove <- Array.append toRemove [|Intersection x y|]) toAdd
            toAdd<-Array.append toAdd work) inst
    (Array.map(Volume) toAdd |> Array.sum) - (Array.map(Volume) toRemove |> Array.sum)

[<EntryPoint>]
let main argv =
    let instructions = File.ReadAllLines(argv.[0]) |> Array.map Process
    printfn "Part 1: %d" <| Activate (Array.filter(IsInit) instructions)
    printfn "Part 2: %d" <| Activate instructions
    0
